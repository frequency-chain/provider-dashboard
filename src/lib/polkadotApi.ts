import type { MsaInfo } from '$lib/storeTypes';
import type { KeyringPair } from '@polkadot/keyring/types';
import type { ChainProperties } from '@polkadot/types/interfaces';
import type { DotApi } from '$lib/storeTypes';
import { ApiPromise, Keyring, WsProvider } from '@polkadot/api';
import { options } from '@frequency-chain/api-augment';

export type AccountMap = Record<string, KeyringPair>;

export async function createApi(networkEndpoint: string): Promise<DotApi> {
  const wsProvider = new WsProvider(networkEndpoint);

  const apiPromise = await ApiPromise.create({
    provider: wsProvider,
    throwOnConnect: true,
    throwOnUnknown: true,
    ...options,
  });

  await apiPromise.isReady;

  const initializedDotApi: DotApi = {
    wsProvider: wsProvider,
    api: apiPromise,
    keyring: new Keyring(),
    selectedEndpoint: networkEndpoint,
    options,
  };
  return initializedDotApi;
}

export function getToken(chain: ChainProperties) {
  const rawUnit = chain.tokenSymbol.toString();
  return rawUnit.slice(1, rawUnit.length - 1);
}

export interface AccountBalances {
  transferable: bigint;
  locked: bigint;
  total: bigint;
}
export async function getBalances(apiPromise: ApiPromise, accountId: string): Promise<AccountBalances> {
  const accountData = (await apiPromise.query.system.account(accountId)).data;
  const free = accountData.free.toBigInt();
  const locked = accountData.frozen.toBigInt();
  const transferable = free - locked;
  const total = free + accountData.reserved.toBigInt();
  return {
    transferable,
    locked,
    total,
  };
}

export async function getMsaInfo(apiPromise: ApiPromise, publicKey: string): Promise<MsaInfo> {
  const received = (await apiPromise?.query.msa.publicKeyToMsaId(publicKey))?.unwrapOrDefault();
  const msaInfo: MsaInfo = { isProvider: false, msaId: 0, providerName: '' };
  msaInfo.msaId = received?.toNumber();
  if (msaInfo.msaId > 0) {
    const providerRegistry = await apiPromise.query.msa.providerToRegistryEntry(msaInfo.msaId);
    if (providerRegistry.isSome) {
      msaInfo.isProvider = true;
      const registryEntry = providerRegistry.unwrap();
      msaInfo.providerName = registryEntry.providerName.toString();
    }
  }
  return msaInfo;
}

export interface CapacityDetails {
  remainingCapacity: bigint;
  totalTokensStaked: bigint;
  totalCapacityIssued: bigint;
  lastReplenishedEpoch: bigint;
}

export async function getCapacityInfo(apiPromise: ApiPromise, msaId: number): Promise<CapacityDetails> {
  let capacityDetails: CapacityDetails = {
    remainingCapacity: 0n,
    totalCapacityIssued: 0n,
    totalTokensStaked: 0n,
    lastReplenishedEpoch: 0n,
  };

  const providerRegistry = await apiPromise.query.msa.providerToRegistryEntry(msaId);

  if (providerRegistry.isSome) {
    const details = (await apiPromise.query.capacity.capacityLedger(msaId)).unwrapOrDefault();
    capacityDetails = {
      remainingCapacity: details.remainingCapacity.toBigInt(),
      totalTokensStaked: details.totalTokensStaked.toBigInt(),
      totalCapacityIssued: details.totalCapacityIssued.toBigInt(),
      lastReplenishedEpoch: details.lastReplenishedEpoch.toBigInt(),
    };
  }

  return capacityDetails;
}
