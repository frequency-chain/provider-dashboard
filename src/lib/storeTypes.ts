import type { ApiPromise, Keyring, WsProvider } from '@polkadot/api';
import type { ApiOptions } from '@polkadot/api/types';

export interface DotApi {
  api?: ApiPromise;
  wsProvider?: WsProvider;
  keyring?: Keyring;
  options?: ApiOptions;
  selectedEndpoint?: string;
}

export const defaultDotApi: DotApi = {
  selectedEndpoint: '',
  options: {},
};

export interface MsaInfo {
  isProvider: boolean;
  msaId: number;
  providerName: string;
}

export interface ChainInfo {
  blockNumber: bigint;
  epochNumber: bigint;
  connected: boolean;
  token: string;
}

export enum TxnStatus {
  LOADING = 'LOADING',
  SUCCESS = 'SUCCESS',
  FAILURE = 'FAILURE',
}

export interface Activity {
  txnId: string;
  txnStatusItems: string[];
  txnStatus: TxnStatus;
}

export type OnChangeFn<T> = (value: T | undefined) => void;