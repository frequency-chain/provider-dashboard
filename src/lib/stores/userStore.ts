import type { Account } from '$lib/stores/accountsStore';
import { storable } from './storable';

export const user = storable<Account>('user', {
  address: '',
  isProvider: false,
  balances: { transferable: 0n, locked: 0n, total: 0n },
});
