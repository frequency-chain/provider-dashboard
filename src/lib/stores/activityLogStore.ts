import { TxnStatus, type Activity } from '$lib/storeTypes';
import type { EventRecord, ExtrinsicStatus } from '@polkadot/types/interfaces';
import type { ISubmittableResult } from '@polkadot/types/types';
import { readonly } from 'svelte/store';
import { storable } from './storable';

//writableActivityLog: Oldest to Newest
const writableActivityLog = storable<Activity[]>('ActivityLog', []);
export const activityLog = readonly(writableActivityLog);

export const handleResult = async (result: ISubmittableResult) => {
  const txnId = result.txHash.toString();
  const activity: Activity = parseActivity(txnId, result);
  addNewTxnStatus(activity);
};

export const handleTxnError = (txnId: string, errorMsg: string) => {
  addNewTxnStatus({ txnId, txnStatusItems: [`${errorMsg}`], txnStatus: TxnStatus.FAILURE });
};

const addNewTxnStatus = (activity: Activity) => {
  writableActivityLog.update((activities: Activity[]) => {
    const curActivity = activities.find((item) => item.txnId === activity.txnId);
    if (curActivity) {
      curActivity.txnStatus = activity.txnStatus;
      curActivity.txnStatusItems = curActivity.txnStatusItems.concat(activity.txnStatusItems);
    } else {
      activities = [activity, ...activities];
    }
    return activities;
  });
};

// how to display the transaction status as it is updated
export function parseActivity(
  txnId: string,
  {
    events = [],
    status,
  }: {
    events?: EventRecord[];
    status: ExtrinsicStatus;
  }
): Activity {
  const txnStatusItems = [];
  let txnStatus: Activity['txnStatus'] = TxnStatus.LOADING;
  try {
    if (status.isInvalid) {
      txnStatusItems.push('Invalid transaction');
    } else if (status.isBroadcast) {
      txnStatusItems.push(`Transaction is Broadcast`);
    } else if (status.isFinalized) {
      txnStatusItems.push(`Transaction is finalized in block hash ${status.asFinalized.toHex()}`);
      events.forEach(({ event }) => {
        if (event.method === 'ExtrinsicSuccess') {
          txnStatusItems.push('Transaction succeeded');
          txnStatus = TxnStatus.SUCCESS;
        } else if (event.method === 'ExtrinsicFailed') {
          txnStatusItems.push('Transaction failed. See chain explorer for details.');
          txnStatus = TxnStatus.FAILURE;
        }
      });
    } else if (status.isInBlock) {
      txnStatusItems.push(`Transaction is included in blockHash ${status.asInBlock.toHex()}`);
    } else {
      console.debug({ status });
    }
  } catch (e: unknown) {
    txnStatusItems.push('Error: ' + (e as Error).toString());
    txnStatus = TxnStatus.FAILURE;
  }
  return { txnStatusItems, txnStatus, txnId };
}

export const clearLog = () => {
  writableActivityLog.set([]);
};
