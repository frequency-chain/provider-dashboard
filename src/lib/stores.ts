import { writable } from 'svelte/store';
import { defaultDotApi, type DotApi } from '$lib/storeTypes';
import { storable } from './stores/storable';
import { derived } from 'svelte/store';
import { user } from './stores/userStore';
import { pageContent } from './stores/pageContentStore';
import { clearLog } from './stores/activityLogStore';
import { createApi } from './polkadotApi';

export const dotApi = writable<DotApi>(defaultDotApi);

export const isLoggedIn = storable<boolean>('isLoggedIn', false);

export const logInPromise = derived([user], ([$user]) =>
  (async () => {
    if ($user?.network?.endpoint) {
      dotApi.set(await createApi($user?.network?.endpoint));
      if ($user.isProvider) {
        isLoggedIn.set(true);
        pageContent.dashboard();
      }
    }
  })()
);

export const logout = () => {
  isLoggedIn.set(false);
  user.set({ address: '', isProvider: false });
  dotApi.set(defaultDotApi);
  storeChainInfo.set({ connected: false, blockNumber: 0n, epochNumber: 0n, token: '' });
  pageContent.login();
  clearLog();
};

export const storeChainInfo = storable('storeChainInfo', {
  connected: false,
  blockNumber: 0n,
  epochNumber: 0n,
  token: '',
});
