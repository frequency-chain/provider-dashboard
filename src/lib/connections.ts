import '@frequency-chain/api-augment';
import type { ApiPromise } from '@polkadot/api/promise';

import type { SubmittableExtrinsic } from '@polkadot/api/promise/types';
import type { InjectedExtension } from '@polkadot/extension-inject/types';
import type { KeyringPair } from '@polkadot/keyring/types';
import type { u64 } from '@polkadot/types';
import type { Signer, SignerPayloadRaw, SignerResult } from '@polkadot/types/types';
import { isFunction, u8aToHex, u8aWrapBytes } from '@polkadot/util';
import { get } from 'svelte/store';
import type { Account } from './stores/accountsStore';
import { handleResult, handleTxnError } from './stores/activityLogStore';
import { user } from './stores/userStore';

interface AddKeyData {
  msaId: string;
  expiration: string;
  newPublicKey: string;
}

export const CENTS: bigint = 1000000n;
export const DOLLARS: bigint = 100n * CENTS;

// No functions in here should have to talk to a component
export async function getBlockNumber(api: ApiPromise): Promise<bigint> {
  if (api && (await api.isReady)) {
    const blockData = await api.rpc.chain.getBlock();
    return blockData.block.header.number.toBigInt();
  }
  return 0n;
}

export async function getEpoch(api: ApiPromise): Promise<bigint> {
  if (api && (await api.isReady)) {
    return ((await api.query.capacity.currentEpoch()) as u64).toBigInt();
  }
  return 0n;
}

// creates the payloads and gets or creates the signatures, then submits the extrinsic
export async function submitAddControlKey(
  api: ApiPromise,
  extension: InjectedExtension | undefined,
  newAccount: Account,
  signingAccount: Account,
  msaId: number
) {
  if (!api || !(await api.isReady)) {
    console.debug('api is not available.');
    return;
  }

  const blockNumber = (await getBlockNumber(api)) as bigint;

  const rawPayload: AddKeyData = {
    msaId: msaId.toString(),
    expiration: (blockNumber + 100n).toString(),
    newPublicKey: newAccount.address,
  };

  const newKeyPayload = api.registry.createType('PalletMsaAddKeyData', rawPayload);

  // mock signatures for fee estimation
  const mockSignature = '0x' + '00'.repeat(64);
  const mockProof = { Sr25519: mockSignature };
  // mock extrinsic for fee estimation
  const mockExtrinsic = api.tx.msa.addPublicKeyToMsa(signingAccount.address, mockProof, mockProof, newKeyPayload);
  // typecheck for testing
  if (typeof mockExtrinsic.paymentInfo === 'function') {
    // Get estimated total cost of txn & user's transferable balance
    const estTotalCost = (await mockExtrinsic.paymentInfo(signingAccount.address)).partialFee.toBigInt();
    const transferable = BigInt(get(user).balances.transferable);
    // Check for adiquite funds
    if (transferable < estTotalCost) throw new Error('User does not have sufficient funds.');
  }

  // Continue with getting real signatures
  const ownerKeySignature = await signPayload(newKeyPayload, signingAccount, extension);
  const newKeySignature = await signPayload(newKeyPayload, newAccount, extension);

  const ownerKeyProof = { Sr25519: ownerKeySignature };
  const newKeyProof = { Sr25519: newKeySignature };
  const extrinsic = api.tx.msa.addPublicKeyToMsa(signingAccount.address, ownerKeyProof, newKeyProof, newKeyPayload);

  submitExtinsic(extrinsic, signingAccount, extension);
}

// creates the payloads and gets or creates the signatures, then submits the extrinsic
export async function submitStake(
  api: ApiPromise,
  extension: InjectedExtension | undefined,
  signingAccount: Account,
  providerId: number,
  stakeAmount: bigint
) {
  if (!api || !(await api.isReady)) {
    console.debug('api is not available.');
    return;
  }

  // Get the estimated transaction fee (does not submit txn)
  const { partialFee } = await api.tx.capacity.stake(providerId, stakeAmount).paymentInfo(signingAccount.address);
  // Get estimated total cost of txn
  const estTotalCost = BigInt(partialFee.toString()) + stakeAmount;
  // Check for adiquite funds
  const hasFunds = BigInt(get(user).balances.transferable) >= estTotalCost;
  if (!hasFunds) throw new Error('User does not have sufficient funds.');

  // Submit txn
  const extrinsic = api.tx.capacity?.stake(providerId, stakeAmount);
  submitExtinsic(extrinsic, signingAccount, extension);
}

// creates the payloads and gets or creates the signatures, then submits the extrinsic
export async function submitUnstake(
  api: ApiPromise,
  extension: InjectedExtension | undefined,
  signingAccount: Account,
  providerId: number,
  unstakeAmount: bigint
) {
  if (!api || !(await api.isReady)) {
    console.debug('api is not available.');
    return;
  }

  // Get the estimated transaction fee (does not submit txn)
  const { partialFee } = await api.tx.capacity.unstake(providerId, unstakeAmount).paymentInfo(signingAccount.address);
  // Get estimated total cost of txn
  const estTotalCost = BigInt(partialFee.toString()) + unstakeAmount;

  // Check for adiquite funds
  const hasFunds = BigInt(get(user).balances.locked) >= estTotalCost;
  if (!hasFunds) throw new Error('User does not have sufficient funds.');

  const extrinsic = api.tx.capacity?.unstake(providerId, unstakeAmount);
  submitExtinsic(extrinsic, signingAccount, extension);
}

function submitExtinsic(
  extrinsic: SubmittableExtrinsic,
  account: Account,
  extension: InjectedExtension | undefined
): Promise<string> {
  if (account.keyringPair) return submitExtrinsicWithKeyring(extrinsic, account.keyringPair);
  if (extension) return submitExtrinsicWithExtension(extension, extrinsic, account.address);
  throw new Error('Unable to find wallet extension');
}

// use the Polkadot extension the user selected to submit the provided extrinsic
async function submitExtrinsicWithExtension(
  extension: InjectedExtension,
  extrinsic: SubmittableExtrinsic,
  signingAddress: string
): Promise<string> {
  // let currentTxDone = false; // eslint-disable-line prefer-const
  try {
    await extrinsic.signAndSend(signingAddress, { signer: extension.signer as Signer, nonce: -1 }, handleResult);
    // await waitFor(() => currentTxDone);
  } catch (e: unknown) {
    const message: string = `${e}`;
    console.debug('caught exception:', message);
    if (message.match(/timeout/i) === null) {
      handleTxnError(extrinsic.hash.toString(), message);
    }
  }
  return extrinsic.hash.toString();
}

// Use the built-in test accounts to submit an extrinsic
async function submitExtrinsicWithKeyring(
  extrinsic: SubmittableExtrinsic,
  signingAccount: KeyringPair
): Promise<string> {
  try {
    await extrinsic.signAndSend(signingAccount, { nonce: -1 }, handleResult);
  } catch (e: unknown) {
    handleTxnError(extrinsic.hash.toString(), `${e}`);
  }
  return extrinsic.hash.toString();
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
async function signPayload(payload: any, account: Account, extension: InjectedExtension | undefined): Promise<string> {
  if (account.keyringPair) return signPayloadWithKeyring(account.keyringPair, payload);
  if (extension) return await signPayloadWithExtension(extension, account.address, payload);
  throw new Error('Unable to find wallet extension');
}

// Use the user's selected Polkadot extension to sign some data
// converting to Sr25519Signature is very important, otherwise the signature length
// is incorrect - just using signature gives:
// Enum(Sr25519):: Expected input with 64 bytes (512 bits), found 15 bytes
export async function signPayloadWithExtension(
  injector: InjectedExtension,
  signingPublicKey: string,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  payload: any
): Promise<string> {
  const signer = injector?.signer;
  let signed: SignerResult;
  if (signer && isFunction(signer.signRaw)) {
    // u8aWrapBytes literally just puts <Bytes></Bytes> around the payload.
    const payloadWrappedToU8a = u8aWrapBytes(payload.toU8a());
    const signerPayloadRaw: SignerPayloadRaw = {
      address: signingPublicKey,
      data: u8aToHex(payloadWrappedToU8a),
      type: 'bytes',
    };
    try {
      signed = await signer.signRaw(signerPayloadRaw);
      return signed?.signature;
    } catch (e: unknown) {
      if (e instanceof Error) {
        return 'ERROR ' + e.message;
      }
    }
  }
  return 'Unknown error';
}

// Use the built-in Alice..Ferdie accounts to sign some data
// returns a properly formatted signature to submit with an extrinsic
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function signPayloadWithKeyring(signingAccount: KeyringPair, payload: any): string {
  try {
    // u8aWrapBytes literally puts <Bytes></Bytes> around the payload.
    return u8aToHex(signingAccount.sign(u8aWrapBytes(payload.toU8a())));
  } catch (e: unknown) {
    if (e instanceof Error) {
      return 'ERROR ' + e.message;
    }
    return 'UNKNOWN ERROR ' + e;
  }
}

//   api: ApiPromise,
//   extension: InjectedExtension | undefined,
//   endpointURL: string,
//   signingAccount: Account,
//   providerName: number,
//   txnStatusCallback: (statusStr: string) => void
//   txnFinishedCallback: () => void
export async function submitCreateProvider(
  api: ApiPromise | undefined,
  extension: InjectedExtension | undefined,
  signingAccount: Account,
  providerName: string
): Promise<string | undefined> {
  if (!api || !(await api.isReady)) {
    console.debug('api is not available.');
    return;
  }

  // Get the estimated txn fee (does not submit txn)
  const estTotalCost = (
    await api.tx.msa.createProvider(providerName).paymentInfo(signingAccount.address)
  ).partialFee.toBigInt();

  // Check for adequate funds
  const hasFunds = BigInt(get(user).balances.transferable) >= estTotalCost;
  if (!hasFunds) throw new Error('User does not have sufficient funds.');

  // Submit txn
  const extrinsic: SubmittableExtrinsic = api.tx.msa.createProvider(providerName);
  return submitExtinsic(extrinsic, signingAccount, extension);
}

export async function submitRequestToBeProvider(
  api: ApiPromise | undefined,
  extension: InjectedExtension | undefined,
  signingAccount: Account,
  providerName: string
): Promise<string | undefined> {
  if (!api || !(await api.isReady)) {
    console.debug('api is not available.');
    return;
  }
  // Get the estimated txn fee (does not submit txn)
  const estTotalCost = (
    await api.tx.msa.proposeToBeProvider(providerName).paymentInfo(signingAccount.address)
  ).partialFee.toBigInt();

  // Check for adequate funds
  const hasFunds = BigInt(get(user).balances.transferable) >= estTotalCost;
  if (!hasFunds) throw new Error('User does not have sufficient funds.');

  // submit txn
  const extrinsic: SubmittableExtrinsic = api.tx.msa.proposeToBeProvider(providerName);
  return submitExtinsic(extrinsic, signingAccount, extension);
}

//Use the current api and signing account to create an MSA for this account
export async function submitCreateMsa(
  api: ApiPromise | undefined,
  extension: InjectedExtension | undefined,
  signingAccount: Account
): Promise<string | undefined> {
  if (!api || !(await api.isReady)) {
    console.debug('api is not available.');
    return;
  }
  // Get the estimated txn fee (does not submit txn)
  const estTotalCost = (await api.tx.msa.create().paymentInfo(signingAccount.address)).partialFee.toBigInt();

  // Check for adequate funds
  const hasFunds = BigInt(get(user).balances.transferable) >= estTotalCost;
  if (!hasFunds) throw new Error('User does not have sufficient funds.');

  const extrinsic: SubmittableExtrinsic = api.tx.msa.create();
  return submitExtinsic(extrinsic, signingAccount, extension);
}
