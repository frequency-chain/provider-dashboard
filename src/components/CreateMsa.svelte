<script lang="ts">
  import { dotApi } from '$lib/stores';
  import { TxnStatus, type Activity, type MsaInfo } from '$lib/storeTypes';
  import type { ApiPromise } from '@polkadot/api';
  import { getExtension } from '$lib/utils';
  import { submitCreateMsa } from '$lib/connections';
  import { user } from '$lib/stores/userStore';
  import { getMsaInfo } from '$lib/polkadotApi';
  import { pageContent } from '$lib/stores/pageContentStore';
  import LoadingIcon from '$lib/assets/LoadingIcon.svelte';
  import ActivityLogPreviewItem from './ActivityLogPreviewItem.svelte';
  import { activityLog } from '$lib/stores/activityLogStore';
  import BackToRootButton from '$components/BackHomeButton.svelte';

  let recentActivityItem: Activity | undefined;
  let recentTxnId: Activity['txnId'] | undefined;
  let isInProgress = false;

  // a callback for when a transaction hits a final state
  let createMsaTxnFinished = async (succeeded: boolean) => {
    if (succeeded) {
      const apiPromise = $dotApi.api as ApiPromise;
      const msaInfo: MsaInfo = await getMsaInfo(apiPromise, $user.address);
      isInProgress = false;
      setTimeout(() => {
        $user.msaId = msaInfo.msaId;
      }, 1500);
      return;
    }
    isInProgress = false;
  };

  const checkIsFinished = async () => {
    if (recentActivityItem && recentActivityItem.txnStatus !== TxnStatus.LOADING) {
      await createMsaTxnFinished(recentActivityItem.txnStatus === TxnStatus.SUCCESS);
    }
  };

  $: {
    recentActivityItem = $activityLog.find((value) => value.txnId === recentTxnId);
    checkIsFinished();
  }

  const doCreateMsa = async (_evt: Event) => {
    isInProgress = true;
    recentTxnId = await submitCreateMsa($dotApi.api, await getExtension($user), $user);
    recentActivityItem = $activityLog.find((value) => value.txnId === recentTxnId);
  };
</script>

<div id="create-msa" class="flex flex-col gap-3.5 text-sm">
  <div class="label">Create MSA Id</div>
  <p>
    An MSA (Message Source Account) is required to become a provider. This action will create an MSA Id that is
    controlled by the selected Transaction Signing Address above. It is available only on Frequency Testnet.
  </p>
  <form class="flex w-[350px] items-end justify-between">
    <button id="create-msa-btn" on:click|preventDefault={doCreateMsa} disabled={isInProgress} class="btn-primary">
      {#if isInProgress}
        <LoadingIcon />
      {:else}
        Create an MSA
      {/if}
    </button>
    <BackToRootButton/>
  </form>
</div>
{#if recentActivityItem}
  <ActivityLogPreviewItem activity={recentActivityItem} />
{/if}
