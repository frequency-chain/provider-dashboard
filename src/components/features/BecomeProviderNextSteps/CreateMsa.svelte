<script lang="ts">
  import { Button } from '@frequency-chain/style-guide';
  import { dotApi } from '$lib/stores';
  import { TxnStatus, type Activity, type MsaInfo } from '$lib/storeTypes';
  import type { ApiPromise } from '@polkadot/api';
  import { getExtension } from '$lib/utils';
  import { submitCreateMsa } from '$lib/connections';
  import { user } from '$lib/stores/userStore';
  import { getMsaInfo } from '$lib/polkadotApi';
  import LoadingIcon from '$lib/assets/LoadingIcon.svelte';
  import { activityLog } from '$lib/stores/activityLogStore';
  import BackToRootButton from '$atoms/BackHomeButton.svelte';
  import ActivityLogPreviewItem from '$features/ActivityLogItem/ActivityLogItem.svelte';

  interface Props {
    isLoading: boolean;
  }

  let { isLoading = $bindable(false) }: Props = $props();

  let recentActivityItem: Activity | undefined = $state();
  let recentTxnId: Activity['txnId'] | undefined = $state();
  let error: string | null = $state(null);

  // a callback for when a transaction hits a final state
  let createMsaTxnFinished = async (succeeded: boolean) => {
    if (succeeded) {
      const apiPromise = $dotApi.api as ApiPromise;
      const msaInfo: MsaInfo = await getMsaInfo(apiPromise, $user.address);
      isLoading = false;
      setTimeout(() => {
        $user.msaId = msaInfo.msaId;
      }, 1500);
      return;
    }
    isLoading = false;
  };

  const checkIsFinished = async () => {
    if (recentActivityItem && recentActivityItem.txnStatus !== TxnStatus.LOADING) {
      await createMsaTxnFinished(recentActivityItem.txnStatus === TxnStatus.SUCCESS);
    }
  };

  $effect(() => {
    recentActivityItem = $activityLog.find((value) => value.txnId === recentTxnId);
    checkIsFinished();
  });

  const doCreateMsa = async (_evt: Event) => {
    isLoading = true;
    try {
      recentTxnId = await submitCreateMsa($dotApi.api, await getExtension($user), $user);
    } catch (err) {
      error = (err as Error).message;
    } finally {
      isLoading = false;
    }
    recentActivityItem = $activityLog.find((value) => value.txnId === recentTxnId);
  };
</script>

<div id="create-msa">
  <div class="form-item-label">Create MSA Id</div>
  <p class="smText">
    An MSA (Message Source Account) is required to become a provider. This action will create an MSA Id that is
    controlled by the selected Transaction Signing Address above.
  </p>
  {#if error}<div class="text-error form-item-label">{error}</div>{/if}
  <div class="mt-f16 flex w-[350px] items-end justify-between">
    <Button onclick={doCreateMsa} disabled={isLoading}>
      {#if isLoading}
        <LoadingIcon />
      {:else}
        Create an MSA
      {/if}
    </Button>
    <BackToRootButton />
  </div>
</div>

{#if recentActivityItem}
  <ActivityLogPreviewItem activity={recentActivityItem} />
{/if}
