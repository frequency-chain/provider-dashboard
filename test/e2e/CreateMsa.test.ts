// CreateMsa.test.ts
import { fireEvent, render, screen, waitFor } from '@testing-library/svelte';
import { vi } from 'vitest';
import { writableActivityLog } from '../../src/lib/stores/activityLogStore';
import { user } from '../../src/lib/stores/userStore';

vi.mock('$lib/connections', async () => {
  const actual = await vi.importActual<typeof import('../../src/lib/connections')>('../../src/lib/connections');
  return {
    ...actual,
    submitCreateMsa: vi.fn(),
  };
});

import CreateMsa from '../../src/components/features/BecomeProviderNextSteps/CreateMsa.svelte';
import { submitCreateMsa } from '../../src/lib/connections';
import { TxnStatus } from '../../src/lib/storeTypes';

vi.mock('$lib/utils', async () => {
  const actual = await vi.importActual<typeof import('../../src/lib/utils')>('../../src/lib/utils');
  return {
    ...actual,
    getExtension: actual.getExtension,
    cn: actual.cn,
  };
});

describe('CreateMsa', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    writableActivityLog.set([]);
    user.set({ address: '0x123', msaId: null });
  });

  it('renders with heading and description', () => {
    render(CreateMsa, { isLoading: false });

    expect(screen.getByText('Create MSA Id')).toBeDefined();
    expect(screen.getByText(/An MSA \(Message Source Account\) is required/)).toBeDefined();
    expect(screen.getByRole('button', { name: /Create an MSA/i })).toBeDefined();
  });

  it('disables button and shows loading icon when loading', async () => {
    render(CreateMsa, { isLoading: true });

    const btn = screen.getByRole('button');
    // should have disabled styling
    expect(btn).toMatchSnapshot();
    expect(screen.getByTestId('loading-icon')).toBeDefined();
  });

  it('calls submitCreateMsa on click', async () => {
    render(CreateMsa);
    const btn = screen.getByRole('button', { name: /Create an MSA/i });

    await fireEvent.click(btn);

    await waitFor(() => {
      expect(submitCreateMsa).toHaveBeenCalled();
      expect(screen.getByTestId('loading-icon')).toBeDefined();
    });
  });

  it('shows error message if submitCreateMsa throws', async () => {
    (submitCreateMsa as jest.Mock).mockRejectedValueOnce(new Error('test error'));

    render(CreateMsa);

    const btn = screen.getByRole('button', { name: /Create an MSA/i });
    await fireEvent.click(btn);

    expect(await screen.findByText('test error')).toBeDefined();
  });

  it('renders ActivityLogPreviewItem if recentActivityItem exists', async () => {
    writableActivityLog.set([{ txnId: 'txn-123', txnStatus: TxnStatus.SUCCESS }]);

    render(CreateMsa);

    await fireEvent.click(screen.getByRole('button', { name: /Create an MSA/i }));
    (submitCreateMsa as jest.Mock).mockRejectedValueOnce('0x123456');

    console.log('***screen', screen);

    expect(await screen.findByText(/Hash:/i)).toBeDefined();
  });
});
