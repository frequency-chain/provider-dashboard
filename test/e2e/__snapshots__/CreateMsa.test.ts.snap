// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateMsa > disables button and shows loading icon when loading 1`] = `
<button
  class="h-[46px] px-f24 rounded-full text-center transition-all text-nowrap hover:bg-primary w-[263px] bg-gray3 text-gray2 cursor-default pointer-events-none"
  disabled=""
>
  <!---->
  <div
    class=" font-bold	 gap-f8 text-normal flex items-center justify-center"
  >
    <!---->
    <!---->
    <!---->
    <svg
      class="h-5 w-5 min-w-5 animate-spin"
      data-testid="loading-icon"
      fill="none"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        class="opacity-25"
        cx="12"
        cy="12"
        r="10"
        stroke="currentColor"
        stroke-width="4"
      />
      <path
        class="opacity-75"
        d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
        fill="currentColor"
      />
    </svg>
    
    
    
    
  </div>
  
  <!---->
</button>
`;
